src/Http/Middleware/AuthenticateSession.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+namespace Laravel\Sanctum\Http\Middleware;
+
+use Closure;
+use Illuminate\Auth\AuthenticationException;
+use Illuminate\Auth\SessionGuard;
+use Illuminate\Contracts\Auth\Factory as AuthFactory;
+use Illuminate\Http\Request;
+use Illuminate\Support\Arr;
+use Illuminate\Support\Collection;
+use Symfony\Component\HttpFoundation\Response;
+
+class AuthenticateSession
+{
+    /**
+     * The authentication factory implementation.
+     *
+     * @var \Illuminate\Contracts\Auth\Factory
+     */
+    protected $auth;
+
+    /**
+     * Create a new middleware instance.
+     *
+     * @param  \Illuminate\Contracts\Auth\Factory  $auth
+     * @return void
+     */
+    public function __construct(AuthFactory $auth)
+    {
+        $this->auth = $auth;
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Closure(\Illuminate\Http\Request): (\Symfony\Component\HttpFoundation\Response)  $next
+     *
+     * @throws \Illuminate\Auth\AuthenticationException
+     */
+    public function handle(Request $request, Closure $next): Response
+    {
+        if ($request->hasSession()) {
+            $guards = Collection::make(Arr::wrap(config('sanctum.guard')))
+                ->mapWithKeys(fn ($guard) => [$guard => $this->auth->guard($guard)])
+                ->filter(fn ($guard) => $guard instanceof SessionGuard);
+
+            $shouldLoggedOut = $guards->filter(fn ($guard, $driver) => $request->session()->has('password_hash_'.$driver))
+                ->filter(fn ($quard, $driver) => $request->session()->get('password_hash_'.$driver) !== $request->user()->getAuthPassword());
+
+            if ($shouldLoggedOut->isNotEmpty()) {
+                $shouldLoggedOut->each->logoutCurrentDevice();
+
+                $request->session()->flush();
+
+                throw new AuthenticationException('Unauthenticated.', [...$shouldLoggedOut->keys()->all(), 'sanctum']);
+            }
+
+            $this->storePasswordHashInSession($request, $guards->keys()->first());
+        }
+
+        return $next($request);
```

Comment:
The logging out issue in the failed test, is why the tap is necessary here. Just like `AuthenticateSession`.

The new password needs to be stored after it has been changed.

Reply:
Fixed